File: day02/main.py
def walk(board):
    moves = {'L': (-1, 0), 'R': (1, 0), 'U': (0, -1), 'D': (0, 1)}
    answer = ''
    for line in open('input.txt'):
        y = next(i for i, row in enumerate(board) if '5' in row)
        x = board[y].index('5')
        for ch in line.strip():
            dx, dy = moves[ch]
            nx = min(max(x + dx, 0), len(board[0])-1)
            ny = min(max(y + dy, 0), len(board)-1)
            if board[ny][nx] != ' ':
                x, y = nx, ny
        answer += board[y][x]
    return answer

board1 = ['123', '456', '789']
print(walk(board1))

board2 = ['  1  ',' 234 ','56789',' ABC ','  D  ']
print(walk(board2))

File: day02/main.nim
import strutils, tables

type Pos = tuple[x: int, y: int]
var moves = {'L': (x: -1, y: 0), 'R': (x: 1, y: 0), 'U': (x: 0, y: -1), 'D': (x: 0, y: 1)}.toTable

proc walk(board: seq[string]): string =
  var answer = ""
  var start: Pos
  for y, line in pairs board:
    start.x = line.find "5"
    if start.x != -1:
      start.y = y
      break

  for line in lines "input.txt":
    var pos = start
    for ch in line:
      var d = moves[ch]
      var nx = pos.x + d.x
      var ny = pos.y + d.y
      if nx >= 0 and ny >= 0 and ny < len(board) and nx < len(board[0]) and board[ny][nx] != ' ':
        pos = (x: nx, y: ny)

    add answer, board[pos.y][pos.x]
  answer

let board1 = @["123", "456", "789"]
echo "Answer #1: ", walk(board1)

let board2 = @["  1  "," 234 ","56789"," ABC ","  D  "]
echo "Answer #2: ", walk(board2)
